examples/cis2-dynamic-nft/src/lib.rs

```diff
@@ -0,0 +1,1450 @@
+    //     self.token_metadata_current_state_counter += 1;
+    // }
+
+    fn add_metadata(&mut self, metadata: MetadataUrl) {
+        self.token_metadata_list.push(metadata.clone());
+    }
+}
+/// The contract state,
+/// Note: The specification does not specify how to structure the contract state
+/// and this could be structured in a more space efficient way.
+#[derive(Serial, DeserialWithState, StateClone)]
+#[concordium(state_parameter = "S")]
+struct State<S> {
+    /// The state of addresses.
+    state:        StateMap<Address, AddressState<S>, S>,
+    /// All of the token IDs
+    /// token-metadata-state
+    tokens:       StateMap<ContractTokenId, TokenMetadataState, S>,
+    /// Map with contract addresses providing implementations of additional
+    /// standards.
+    implementors: StateMap<StandardIdentifierOwned, Vec<ContractAddress>, S>,
+}
+
+/// The different errors the contract can produce.
+#[derive(Serialize, Debug, PartialEq, Eq, Reject, SchemaType)]
+enum CustomContractError {
+    /// Failed parsing the parameter.
+    #[from(ParseError)]
+    ParseParams,
+    /// Failed logging: Log is full.
+    LogFull,
+    /// Failed logging: Log is malformed.
+    LogMalformed,
+    /// Invalid contract name.
+    InvalidContractName,
+    /// Only a smart contract can call this function.
+    ContractOnly,
+    /// Failed to invoke a contract.
+    InvokeContractError,
+    /// Failed to upgrade token because of max metadata counter is reached.
+    UpgradeTokenFailedError,
+}
+
+type ContractError = Cis2Error<CustomContractError>;
+
+type ContractResult<A> = Result<A, ContractError>;
+
+/// Mapping the logging errors to ContractError.
+impl From<LogError> for CustomContractError {
+    fn from(le: LogError) -> Self {
+        match le {
+            LogError::Full => Self::LogFull,
+            LogError::Malformed => Self::LogMalformed,
+        }
+    }
+}
+
+/// Mapping errors related to contract invocations to CustomContractError.
+impl<T> From<CallContractError<T>> for CustomContractError {
+    fn from(_cce: CallContractError<T>) -> Self { Self::InvokeContractError }
+}
+
+/// Mapping CustomContractError to ContractError
+impl From<CustomContractError> for ContractError {
+    fn from(c: CustomContractError) -> Self { Cis2Error::Custom(c) }
+}
+
+impl From<NewReceiveNameError> for CustomContractError {
+    fn from(_: NewReceiveNameError) -> Self { Self::InvalidContractName }
+}
+
+impl From<NewContractNameError> for CustomContractError {
+    fn from(_: NewContractNameError) -> Self { Self::InvalidContractName }
+}
+
+impl<S: HasStateApi> State<S> {
+    /// Construct a state with no tokens
+    fn empty(state_builder: &mut StateBuilder<S>) -> Self {
+        State {
+            state:        state_builder.new_map(),
+            tokens:       state_builder.new_map(),
+            implementors: state_builder.new_map(),
+        }
+    }
+
+    /// Mints an amount of tokens with a given address as the owner.
```

Comment:
Please also mention the tokenMetata list and counter and how they are set by this mint function (this part of the logic is special for this contract so it is worth to be mention).